fem/tmop_pa_h2s.cpp

```diff
@@ -0,0 +1,290 @@
+
+// *****************************************************************************
+template<int T_D1D = 0, int T_Q1D = 0, int T_NBZ = 0>
+static void SetupGradPA_2D(const Vector &xe_,
+                           const int NE,
+                           const Array<double> &w_,
+                           const Array<double> &b_,
+                           const Array<double> &g_,
+                           const DenseMatrix &j_,
+                           Vector &dp_,
+                           const int d1d = 0,
+                           const int q1d = 0)
+{
+   constexpr int DIM = 2;
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+   constexpr int NBZ = T_NBZ ? T_NBZ : 1;
+
+   const auto W = Reshape(w_.Read(), Q1D, Q1D);
+   const auto b = Reshape(b_.Read(), Q1D, D1D);
+   const auto g = Reshape(g_.Read(), Q1D, D1D);
+   const auto J = Reshape(j_.Read(), DIM, DIM);
+   const auto X = Reshape(xe_.Read(), D1D, D1D, DIM, NE);
+   auto dP = Reshape(dp_.Write(), DIM, DIM, DIM, DIM, Q1D, Q1D, NE);
+
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, NBZ,
+   {
+      constexpr int DIM = 2;
+      const int tidz = MFEM_THREAD_ID(z);
+      const int D1D = T_D1D ? T_D1D : d1d;
+      const int Q1D = T_Q1D ? T_Q1D : q1d;
+      constexpr int NBZ = T_NBZ ? T_NBZ : 1;
+      constexpr int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+      constexpr int MD1 = T_D1D ? T_D1D : MAX_D1D;
+
+      MFEM_SHARED double s_BG[2][MQ1*MD1];
+      double (*B)[MD1]  = (double (*)[MD1])(s_BG[0]);
+      double (*G)[MD1]  = (double (*)[MD1])(s_BG[1]);
+
+      MFEM_SHARED double s_X[2][NBZ][MD1*MD1];
+      double (*Xx)[MD1]  = (double (*)[MD1])(s_X[0] + tidz);
+      double (*Xy)[MD1]  = (double (*)[MD1])(s_X[1] + tidz);
+
+      MFEM_SHARED double s_DQ[4][NBZ][MD1*MQ1];
+      double (*XxB)[MQ1] = (double (*)[MQ1])(s_DQ[0] + tidz);
+      double (*XxG)[MQ1] = (double (*)[MQ1])(s_DQ[1] + tidz);
+      double (*XyB)[MQ1] = (double (*)[MQ1])(s_DQ[2] + tidz);
+      double (*XyG)[MQ1] = (double (*)[MQ1])(s_DQ[3] + tidz);
+
+      MFEM_SHARED double s_QQ[4][NBZ][MQ1*MQ1];
+      double (*Xx0)[MQ1] = (double (*)[MQ1])(s_QQ[0] + tidz);
+      double (*Xx1)[MQ1] = (double (*)[MQ1])(s_QQ[1] + tidz);
+      double (*Xy0)[MQ1] = (double (*)[MQ1])(s_QQ[2] + tidz);
+      double (*Xy1)[MQ1] = (double (*)[MQ1])(s_QQ[3] + tidz);
+
+      // Load X(x,y)
+      MFEM_FOREACH_THREAD(dy,y,D1D)
+      {
+         MFEM_FOREACH_THREAD(dx,x,D1D)
+         {
+            Xx[dy][dx] = X(dx,dy,0,e);
+            Xy[dy][dx] = X(dx,dy,1,e);
+         }
+      }
+      // Load B and G matrices
+      if (tidz == 0)
+      {
+         MFEM_FOREACH_THREAD(d,y,D1D)
+         {
+            MFEM_FOREACH_THREAD(q,x,Q1D)
+            {
+               B[q][d] = b(q,d);
+               G[q][d] = g(q,d);
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+      MFEM_FOREACH_THREAD(dy,y,D1D)
+      {
+         MFEM_FOREACH_THREAD(qx,x,Q1D)
+         {
+            double u[2] {};
```

Comment:
Does this set the data to zero too? I thought you had to do ``u[2] = {0};``

Reply:
It does, yes. The `={0}` will set all (first and tail) to zero too, but `={1}` wont (first to 1.0, tail to 0.0), which I think is misleading.

Reply:
For clarity, I'd recommend to just use standard initialization: `double u[2] = {0.0, 0.0};` -- not everyone is familiar with all these strange shortcuts.

Reply:
`done`